datafusion/optimizer/src/push_down_filter.rs

```diff
@@ -998,7 +998,19 @@ impl OptimizerRule for PushDownFilter {
                         filter_predicates.len());
                 }
 
-                let zip = filter_predicates.into_iter().zip(results);
+                let zip =
+                    filter_predicates
+                        .into_iter()
+                        .zip(results)
+                        .map(|(expr, res)| {
+                            let filter_pushdown_type = if expr.is_volatile() {
```

Comment:
shouldn't we check `is_volatile` before calling `scan.source.supports_filters_pushdown(...)`?

Reply:
As I understand, we can ask a table source to tell if the filter is supported even for volatile expressions.

I thought the main difference is which call is more expensive. `supports_filters_pushdown` for different sources often provides blanket all-supported or all-unsupported for all predicates. So we can save `is_volatile` calls by filtering out `Unsupported` values before. In contrast, if we check for volatility first and only then ask the source if the filters are supported, we can waste `is_volatilty` calls. If the `supports_filters_pushdown` is more costly, let's reverse it, of course.

I pushed a change to check the volatility after filtering out unsupported predicates. If we prefer the inverse order, I could refactor this optimiser quite a bit to:
- Get a subset of supported predicates
- Apply the `is_volatile` filter to it
- Build `new_scan_filters` and `new_predicate`

Reply:
If there is possible that any table source can support volatile function then this checking should happen in `supports_filters_pushdown`.

Otherwise, we should check `is_volatile` before `supports_filters_pushdown`, so those table source don't need to care about whether the function is volatile or not.

Given that volatile has no deterministic result, I think it doesn't make sense to pushdown filter with those function, therefore I think we can check volatility beforehand